python/ql/src/experimental/semmle/python/frameworks/Xml.qll

```diff
@@ -8,298 +8,456 @@ private import semmle.python.dataflow.new.DataFlow
-      )
+    override predicate vulnerableTo(XML::XMLVulnerabilityKind kind) {
+      this.getParserArg() = saxParserWithFeatureExternalGesTurnedOn() and
+      (kind.isXxe() or kind.isDtdRetrieval())
       or
-      kind = "XXE" and not exists(this.getArgByName("parser"))
+      (kind.isBillionLaughs() or kind.isQuadraticBlowup())
     }
   }
+}
 
+private module Lxml {
   /**
-   * Gets a call to `xmltodict.parse`.
-   *
-   * Given the following example:
+   * Provides models for `lxml.etree` parsers
    *
-   * ```py
-   * xmltodict.parse(xml_content, disable_entities=False)
-   * ```
-   *
-   * * `this` would be `xmltodict.parse(xml_content, disable_entities=False)`.
-   * * `getAnInput()`'s result would be `xml_content`.
-   * * `vulnerable(kind)`'s `kind` would be `Billion Laughs` and `Quadratic Blowup`.
+   * See https://lxml.de/apidoc/lxml.etree.html?highlight=xmlparser#lxml.etree.XMLParser
    */
-  private class XMLtoDictParsing extends DataFlow::CallCfgNode, XML::XMLParsing::Range {
-    XMLtoDictParsing() { this = API::moduleImport("xmltodict").getMember("parse").getACall() }
+  module XMLParser {
+    /**
+     * A source of instances of `lxml.etree` parsers, extend this class to model new instances.
+     *
+     * This can include instantiations of the class, return values from function
+     * calls, or a special parameter that will be set when functions are called by an external
+     * library.
+     *
+     * Use the predicate `XMLParser::instance()` to get references to instances of `lxml.etree` parsers.
+     */
+    abstract class InstanceSource extends DataFlow::LocalSourceNode {
+      /** Holds if this instance is vulnerable to `kind`. */
+      abstract predicate vulnerableTo(XML::XMLVulnerabilityKind kind);
+    }
 
-    override DataFlow::Node getAnInput() { result = this.getArg(0) }
+    /**
+     * A call to `lxml.etree.XMLParser`.
+     *
+     * See https://lxml.de/apidoc/lxml.etree.html?highlight=xmlparser#lxml.etree.XMLParser
+     */
+    private class LXMLParser extends InstanceSource, DataFlow::CallCfgNode {
+      LXMLParser() {
+        this = API::moduleImport("lxml").getMember("etree").getMember("XMLParser").getACall()
+      }
 
-    override predicate vulnerable(string kind) {
-      kind = ["Billion Laughs", "Quadratic Blowup"] and
-      this.getArgByName("disable_entities").getALocalSource().asExpr() = any(False f)
+      // NOTE: it's not possible to change settings of a parser after constructing it
+      override predicate vulnerableTo(XML::XMLVulnerabilityKind kind) {
+        kind.isXxe() and
+        (
+          // resolve_entities has default True
+          not exists(this.getArgByName("resolve_entities"))
+          or
+          this.getArgByName("resolve_entities").getALocalSource().asExpr() = any(True t)
+        )
+        or
+        (kind.isBillionLaughs() or kind.isQuadraticBlowup()) and
+        this.getArgByName("huge_tree").getALocalSource().asExpr() = any(True t)
```

Comment:
Billion Laughs and Quadratic Blowup relies on entity expansion (not actual resolution, though), but if I remember well from when I tested the lib, `huge_tree` needed `resolve_entities` in order for the former vulnerabilities to be exploitable. 

(Changed in https://github.com/jorgectf/codeql/pull/9/commits/e295399f7096f92592ea7aa4d1286619bf39f8d0)

Reply:
Ohh, I see your point. Yes, `huge_tree` in itself is not enough :+1:

Reply:
Fixed in f0131af